test/enclave_wrangler_tests/test_dataset_upload.py

```diff
@@ -0,0 +1,77 @@
+"""Tests
+
+How to run:
+    python -m unittest discover
+
+TODO's
+ - 1. Test framework: Current implementation is ad-hoc for purposes of development.
+ - 2. Change from validate to apply, or do both
+"""
+import os
+import pickle
+import sys
+import unittest
+from datetime import datetime, timedelta
+from pathlib import Path
+from typing import Dict, List, Union
+
+import pandas as pd
+from requests import Response
+from sqlalchemy.exc import IntegrityError
+
+from enclave_wrangler.dataset_upload import upload_new_cset_container_with_concepts_from_csv, \
+    upload_new_cset_version_with_concepts_from_csv
+
+TEST_DIR = os.path.dirname(__file__)
+PROJECT_ROOT = Path(TEST_DIR).parent
+# todo: why is this necessary in this case and almost never otherwise?
+# https://stackoverflow.com/questions/33862963/python-cant-find-my-module
+sys.path.insert(0, str(PROJECT_ROOT))
+
+TEST_INPUT_DIR = os.path.join(TEST_DIR, '../input', 'test_enclave_wrangler')
+TEST_SCHEMA = 'test_n3c'
+yesterday: str = (datetime.now() - timedelta(days=1)).isoformat() + 'Z'  # works: 2023-01-01T00:00:00.000Z
+
+class TestDatasetUpload(unittest.TestCase):
+    def test_upload_cset_container_from_csv(self):
```

Comment:
This test failed, it looks like a genuine unit test that is failing because the function that it test is not behaving as expected but because of all the other things in these files that don't seem to be working as expected I can't say that with certainty.

Reply:
If you copy/paste the error message I can help more, but I think you are right, based on this:
`PROJECT_ROOT = Path(TEST_DIR).parent`
It needs an extra `.parent` at the end now.

Reply:
I fixed the pathing!

I did not actually run the test to full execution (nor many of the tests), because since many things have been changing, and I haven't kept them up to date, I expect many of the tests to fail. I don't think I mentioned that... so now you know! Don't worry about my old tests passing unless it's something I'm specifically having you work on. You can focus your attention on the tests that you've been developing and will be going forward.

To be honest I'm really not proud of a lot of what I've done in unit testing: (i) tests got stale and broken, (ii) many tests incomplete / commented out, (iii) poor test coverage, (iv) general messiness, (v) haven't been running tests regularly nor set an automated way to do that.